pool/native.go

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package pool
+
+import (
+	"reflect"
+	"unsafe"
+)
+
+// NativePoolOptions specify options for NativePool.
+type NativePoolOptions struct {
+	Construct func(ptr interface{})
+	Size      uint
+	Type      reflect.Type
+}
+
+// NativePool represents an object pool which is opaque to the Go GC.
+type NativePool interface {
+	Get() interface{}
+	GetOr(OverflowFn) interface{}
+	Put(interface{})
+
+	// Owns determines if the object belongs to the pool.
+	Owns(interface{}) bool
+
+	// Size returns the used and the total capacity of the pool.
+	Size() (uint64, uint64)
+}
+
+// OverflowFn produces non-pooled objects.
+type OverflowFn func() interface{}
+
+// NewNativePool constructs a new NativePool.
+func NewNativePool(opts NativePoolOptions) NativePool {
+	p := &nativePool{
+		free: make(chan uint64, opts.Size),
+		opts: opts,
+		// TODO(@kobolog): alignment & padding.
+		step: uint64(hsz + opts.Type.Size())}
+
+	p.size = uint64(p.opts.Size) * p.step
+
+	if p.size == 0 {
+		panic("native-pool: pool size is zero")
+	}
+
+	if p.step%uint64(p.opts.Type.Align()) != 0 {
+		panic("native-pool: invalid alignment")
+	}
+
+	if p.opts.Construct == nil {
+		p.opts.Construct = func(interface{}) {}
+	}
+
+	p.init()
+
+	return p
+}
+
+type nativePool struct {
+	pool       []byte
+	free       chan uint64
+	opts       NativePoolOptions
+	step, size uint64
+}
+
+func (p *nativePool) Size() (uint64, uint64) {
+	return p.size - uint64(len(p.free))*p.step, uint64(cap(p.free)) * p.step
+}
+
+type hdr struct {
+	// Offset from the beginning of the arena to the object.
+	idx uint64
+}
+
+const (
+	hsz = unsafe.Sizeof(*(*hdr)(nil))
```

Comment:
Moved all unexported types together. Won't move it to the top of the file since it's for **exported** stuff, not implementation detail.